File: MDPatientManagement/src/domain/medicalRecordNumber.ts
import { ValueObject } from "../core/domain/ValueObject";
import { Result } from "../core/logic/Result";

interface MedicalRecordNumberProps {
    medicalRecordNumber: string;
}

export class MedicalRecordNumber extends ValueObject<MedicalRecordNumberProps> {

    get value(): string {
        return this.props.medicalRecordNumber;
    }

    public constructor(props: MedicalRecordNumberProps) {
        super(props);
    }

    public static create(props: MedicalRecordNumberProps): Result<MedicalRecordNumber> {

        const medicalRecordNumberDto = props.medicalRecordNumber;

        const year = medicalRecordNumberDto.slice(0, 4);
        const month = medicalRecordNumberDto.slice(4, 6);
        const seqNumber = medicalRecordNumberDto.slice(6);

        if (!!medicalRecordNumberDto === false || medicalRecordNumberDto.length < 10) {
            return Result.fail<MedicalRecordNumber>('Invalid Medical Record Number format.')
        } else if (isNaN(Number(year)) || isNaN(Number(month)) || isNaN(Number(seqNumber))) {
            return Result.fail<MedicalRecordNumber>('Invalid year, month, or sequential number format.');
        } else if (seqNumber.length !== 6) {
            return Result.fail<MedicalRecordNumber>('Sequential number must be a 6-digit number.');
        } else {
            return Result.ok<MedicalRecordNumber>(new MedicalRecordNumber({
                medicalRecordNumber: props.medicalRecordNumber,
            }));
        }
    }

}

File: MDBackoffice/domain/patients/MedicalRecordNumber.cs
using System;
using MDBackoffice.Domain.Shared;

namespace MDBackoffice.Domain.Patients
{
    public class MedicalRecordNumber : EntityId
    {
        public MedicalRecordNumber(string seqNumber, bool isUnfinished)
            : base($"{DateTime.Now.Year}{DateTime.Now.Month.ToString("D2")}{seqNumber}")
        {
            if (seqNumber.Length != 6 || !int.TryParse(seqNumber, out _))
            {
                throw new BusinessRuleValidationException("Sequential number must be a 5-digit number.");
            }
            if (isUnfinished) 
            {
                return;
            }
        }

        public MedicalRecordNumber(string value)
                 : base(value)
        {

        }
        protected override object createFromString(string text)
        {
            if (string.IsNullOrEmpty(text) || text.Length < 10)
            {
                throw new BusinessRuleValidationException("Invalid Medical Record Number format.");
            }

            string year = text[..4];
            string month = text.Substring(4, 2);
            string seqNumber = text[6..];

            if (!int.TryParse(year, out _) || !int.TryParse(month, out _) || !int.TryParse(seqNumber, out _))
            {
                throw new BusinessRuleValidationException("Invalid year, month, or sequential number format.");
            }

            if (seqNumber.Length != 6)
            {
                throw new BusinessRuleValidationException("Sequential number must be a 6-digit number.");
            }

            // Now, create a new instance using the valid seqNumber
            return text;
        }


        public override string AsString()
        {
            return Value;
        }
    }
}